src/v/storage/kvstore.cc

```diff
@@ -575,18 +575,28 @@ ss::future<> kvstore::replay_segments(segment_set segs) {
         co_await (*it)->close();
     }
 
+    // we want to defer local storage effects until we've confirmed that
+    // the logical version is compatible with the rest of the cluster
+    _startup_garbage.reserve(std::distance(match, segs.end()) * 2);
+    for (auto it = match; it != segs.end(); it++) {
+        auto seg = *it;
+        _startup_garbage.emplace_back(seg->reader().path().string());
+        _startup_garbage.emplace_back(seg->index().path().string());
+    }
+}
+ss::future<> kvstore::collect_startup_garbage() {
     // saving a snapshot right after recovery during start-up prevents an
     // accumulation of segments in cases where the system restarts many
     // times without ever filling up a segment and snapshotting when
     // rolling. they'll be removed on the next startup.
     co_await save_snapshot();
 
     // gc the replayed segments now that the snapshot has been taken.
-    for (auto it = match; it != segs.end(); it++) {
-        auto seg = *it;
-        co_await ss::remove_file(seg->reader().path().string());
-        co_await ss::remove_file(seg->index().path().string());
+    for (const auto& fname : _startup_garbage) {
+        co_await ss::remove_file(fname);
     }
+
+    _startup_garbage.clear();
```

Comment:
See: https://en.cppreference.com/w/cpp/container/vector/clear

Capacity is unchanged. If you make this a `ss::chunked_fifo` then clear will remove capacity.

Reply:
Ah, right. 

> `ss::chunked_fifo`

That's a nifty container 🙂